File: base/src/vsdk/toolkit/io/geometry/WriterGts.java
//===========================================================================
//=-------------------------------------------------------------------------=
//= Module history:                                                         =
//= - October 26 2007 - Oscar Chavarro: Original base version               =
//===========================================================================

package vsdk.toolkit.io.geometry;

// Java basic classes
import java.util.ArrayList;
import java.io.OutputStream;

// VSDK Classes
import vsdk.toolkit.common.Matrix4x4;
import vsdk.toolkit.common.Triangle;
import vsdk.toolkit.common.Vertex;
import vsdk.toolkit.common.Vector3D;
import vsdk.toolkit.environment.geometry.Geometry;
import vsdk.toolkit.environment.geometry.TriangleMesh;
import vsdk.toolkit.environment.geometry.TriangleMeshGroup;
import vsdk.toolkit.environment.geometry.FunctionalExplicitSurface;
import vsdk.toolkit.environment.scene.SimpleBody;
import vsdk.toolkit.environment.scene.SimpleScene;
import vsdk.toolkit.io.PersistenceElement;

class _WriterGtsEdge
{
    public int from;
    public int to;
}

class _WriterGtsTriangle
{
    public int p0;
    public int p1;
    public int p2;
}

public class WriterGts extends PersistenceElement {

    private static int
    addEdge(ArrayList<_WriterGtsEdge> edges, int from, int to)
    {
        _WriterGtsEdge e;
        int i;

        for ( i = 0; i < edges.size(); i++ ) {
            e = edges.get(i);
            if ( (e.from == from && e.to == to) ||
                 (e.from == to && e.to == from) ) {
                return i;
            }
        }

        i = edges.size();
        e = new _WriterGtsEdge();
        e.from = from;
        e.to = to;
        edges.add(e);
        return i;
    }

    private static long
    exportMesh(OutputStream inOutputStream, TriangleMesh mesh, long offset)
        throws Exception
    {
        Vertex v[] = mesh.getVertexes();
        Triangle t[] = mesh.getTriangles();
        ArrayList<_WriterGtsEdge> edges;
        ArrayList<_WriterGtsTriangle> triangles;
        edges = new ArrayList<_WriterGtsEdge>();
        triangles = new ArrayList<_WriterGtsTriangle>();
        _WriterGtsEdge e;
        _WriterGtsTriangle tt;
        int i;

        for ( i = 0; i < t.length; i++ ) {
            tt = new _WriterGtsTriangle();
            tt.p0 = addEdge(edges, t[i].p0, t[i].p1);
            tt.p1 = addEdge(edges, t[i].p1, t[i].p2);
            tt.p2 = addEdge(edges, t[i].p2, t[i].p0);
            triangles.add(tt);
        }

        writeAsciiLine(inOutputStream, "" + v.length + " " + edges.size() + " " + triangles.size() + " GtsSurface GtsFace GtsEdge GtsVertex");

        for ( i = 0; i < v.length; i++ ) {
            writeAsciiLine(inOutputStream, "" + v[i].position.x + " " +
                v[i].position.y + " " + v[i].position.z);
        }
        for ( i = 0; i < edges.size(); i++ ) {
            e = edges.get(i);
            writeAsciiLine(inOutputStream, "" + (e.from+1) + " " + (e.to+1));
        }
        for ( i = 0; i < triangles.size(); i++ ) {
            tt = triangles.get(i);
            writeAsciiLine(inOutputStream, "" + (tt.p0+1) + " " + (tt.p1+1) + " " + (tt.p2+1));
        }

        return offset + v.length;
    }

    public static void
    exportEnvironment(OutputStream inOutputStream, SimpleScene inScene)
        throws Exception
    {
        //-----------------------------------------------------------------
        ArrayList<SimpleBody> objs;
        Geometry g;
        TriangleMesh mesh;

        objs = inScene.getSimpleBodies();
        long baseVertexStart = 0;

        int i;
        for ( i = 0; i < objs.size(); i++ ) {

            //-----------------------------------------------------------------
            g = objs.get(i).getGeometry();
            mesh = null;
            if ( g instanceof FunctionalExplicitSurface ) {
                mesh = ((FunctionalExplicitSurface)g).getInternalTriangleMesh();
            }
            else if ( g instanceof TriangleMesh ) {
                mesh = (TriangleMesh)g;
            }

            //-----------------------------------------------------------------
            if ( mesh != null ) {
                baseVertexStart += exportMesh(inOutputStream, mesh, baseVertexStart);
            }
        }
    }
}

//===========================================================================
//= EOF                                                                     =
//===========================================================================

File: cpp/base/src/main/vsdk/toolkit/io/geometry/WriterGts.cpp
#include "java/lang/Double.h"
#include "java/util/ArrayList.txx"
#include "vsdk/toolkit/environment/scene/SimpleBody.h"
#include "vsdk/toolkit/environment/scene/SimpleScene.h"
#include "vsdk/toolkit/io/geometry/WriterCommon.h"
#include "vsdk/toolkit/io/geometry/WriterGts.h"

namespace {

class WriterGtsEdge {
public:
    int from;
    int to;
    WriterGtsEdge() : from(0), to(0) {}
};

class WriterGtsTriangle {
public:
    int point0;
    int point1;
    int point2;
    WriterGtsTriangle() : point0(0), point1(0), point2(0) {}
};

void line(java::OutputStream& os, const java::String& text)
{
    PersistenceElement::writeAsciiLine(os, text.c_str());
}

int addEdge(java::ArrayList<WriterGtsEdge>& edges, int from, int to)
{
    int i;

    for ( i = 0; i < edges.size(); i++ ) {
        const WriterGtsEdge& e = edges[i];
        if ( (e.from == from && e.to == to) ||
             (e.from == to && e.to == from) ) {
            return i;
        }
    }

    i = (int)edges.size();
    WriterGtsEdge e;
    e.from = from;
    e.to = to;
    edges.add(e);
    return i;
}

long long exportMesh(java::OutputStream& os, TriangleMesh* mesh,
                     long long offset)
{
    int nv = mesh->getNumVertices();
    int nt = mesh->getNumTriangles();
    java::ArrayList<WriterGtsEdge> edges;
    java::ArrayList<WriterGtsTriangle> triangles;
    int i;

    //- Compute edges -------------------------------------------------
    java::ArrayList<int>& t = mesh->getTriangleIndexes();
    for ( i = 0; i < nt; i++ ) {
        WriterGtsTriangle tt;
        tt.point0 = addEdge(edges, t[3*i], t[3*i+1]);
        tt.point1 = addEdge(edges, t[3*i+1], t[3*i+2]);
        tt.point2 = addEdge(edges, t[3*i+2], t[3*i]);
        triangles.add(tt);
    }

    //- Write GTS header ----------------------------------------------
    line(os, java::String::valueOf(nv) + " " +
         java::String::valueOf((long)edges.size()) + " " +
         java::String::valueOf((long)triangles.size()) +
         " GtsSurface GtsFace GtsEdge GtsVertex");

    //- Write vertices ------------------------------------------------
    java::ArrayList<double>& v = mesh->getVertexPositions();
    for ( i = 0; i < nv; i++ ) {
        line(os, java::Double::toString(v[3*i]) + " " +
             java::Double::toString(v[3*i+1]) + " " +
             java::Double::toString(v[3*i+2]));
    }

    //- Write edges ---------------------------------------------------
    for ( i = 0; i < edges.size(); i++ ) {
        const WriterGtsEdge& e = edges[i];
        line(os, java::String::valueOf(e.from+1) + " " +
             java::String::valueOf(e.to+1));
    }

    //- Write triangles -----------------------------------------------
    for ( i = 0; i < triangles.size(); i++ ) {
        const WriterGtsTriangle& tt = triangles[i];
        line(os, java::String::valueOf(tt.point0+1) + " " +
             java::String::valueOf(tt.point1+1) + " " +
             java::String::valueOf(tt.point2+1));
    }

    return offset + nv;
}

}

void WriterGts::exportEnvironment(java::OutputStream& outputStream,
                                  SimpleScene* scene)
{
    java::ArrayList<SimpleBody*>& objs = scene->getSimpleBodies();
    long long baseVertexStart = 0;
    int i;

    for ( i = 0; i < objs.size(); i++ ) {
        TriangleMesh* mesh = writerMeshOf(objs.get(i)->getGeometry());

        //-----------------------------------------------------------------
        if ( mesh != nullptr ) {
            baseVertexStart += exportMesh(outputStream, mesh, baseVertexStart);
        }
    }
}
